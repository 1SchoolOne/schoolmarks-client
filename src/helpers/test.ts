import { MockedObject } from 'vitest'

import { axiosInstance } from '@api/axiosInstance'

vi.mock('../api/axiosInstance')
/**
 * Permet de mocker les requÃªtes axios.
 *
 * @example
 *
 * mockAxios.get.mockResolvedValue({ data: [{userId: 1, first_name: 'toto' }] })
 */
export const mockAxios = axiosInstance as MockedObject<typeof axiosInstance>
