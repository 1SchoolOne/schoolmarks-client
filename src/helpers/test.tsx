import { PropsWithChildren, ThemeProvider } from '@1schoolone/ui'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { RenderOptions, render, render as renderReact } from '@testing-library/react'
import { ReactNode, useMemo } from 'react'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'

import { IdentityProvider } from '@contexts'

const testingQueryClient = new QueryClient({
	defaultOptions: { queries: { staleTime: 0, retry: false } },
})

function AllProviders({ children }: PropsWithChildren) {
	const testingRouter = useMemo(
		() =>
			createMemoryRouter(
				[{ path: '/:classSessionId', element: <IdentityProvider>{children}</IdentityProvider> }],
				{ initialEntries: ['/123'] },
			),
		[children],
	)

	return (
		<ThemeProvider>
			<RouterProvider router={testingRouter} />
		</ThemeProvider>
	)
}

function renderWithQueryClient(
	client: QueryClient,
	ui: React.ReactElement,
	options?: RenderOptions,
): ReturnType<typeof renderReact> {
	const { rerender, ...result } = renderReact(
		<QueryClientProvider client={client}>{ui}</QueryClientProvider>,
		options,
	)

	return {
		...result,
		rerender: (rerenderUi: React.ReactNode) =>
			rerender(<QueryClientProvider client={client}>{rerenderUi}</QueryClientProvider>),
	}
}

/** This helper will render the React node only with the ThemeProvider. */
export function renderWithThemeProvider(
	elements: ReactNode,
	options?: Omit<RenderOptions, 'wrapper'>,
) {
	return render(elements, { ...options, wrapper: ThemeProvider })
}

/**
 * This helper will render the React node with all the following providers:
 *
 * - ThemeProvider
 * - RouterProvider
 * - QueryClientProvider
 */
export function renderWithProviders(elements: ReactNode, options?: RenderOptions) {
	return renderWithQueryClient(testingQueryClient, <AllProviders>{elements}</AllProviders>, options)
}
