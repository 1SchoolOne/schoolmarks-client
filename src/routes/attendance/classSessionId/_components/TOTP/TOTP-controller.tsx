import { useQuery } from '@tanstack/react-query'
import axios from 'axios'
import dayjs from 'dayjs'
import { useEffect, useRef, useState } from 'react'
import { useLoaderData, useParams } from 'react-router-dom'

import { AXIOS_DEFAULT_CONFIG } from '@api/axios'
import { getClassSessionQueryOptions } from '@api/classSessions'

import { classSessionloader } from '../../..'

export function useController() {
	const initialData = useLoaderData() as Awaited<ReturnType<typeof classSessionloader>>
	const params = useParams()
	const [totpCountdown, setTotpCountdown] = useState(0)
	const [totpExpiresAt, setTotpExpiresAt] = useState(dayjs().add(15, 'seconds'))
	const previousTotp = useRef<string>()

	const classSessionQueryOptions = getClassSessionQueryOptions(params.classSessionId)

	const { data: classSession } = useQuery({
		...classSessionQueryOptions,
		initialData,
		enabled: typeof params.classSessionId === 'string',
	})

	const { data: totp } = useQuery({
		queryKey: ['checkin-session', 'totp', classSession?.checkin_session?.id],
		queryFn: async () => {
			const { data } = await axios.get<{ totp: string }>(
				`/checkin_sessions/${classSession?.checkin_session?.id}/totp/`,
				AXIOS_DEFAULT_CONFIG,
			)

			if (previousTotp.current !== data.totp) {
				setTotpExpiresAt(dayjs().add(15.5, 'seconds'))
			}

			previousTotp.current = data.totp

			return data.totp
		},
		refetchIntervalInBackground: true,
		refetchInterval: 1000,
		staleTime: 1000,
		enabled: !!classSession?.checkin_session,
	})

	useEffect(
		function updateCountdown() {
			const interval = setInterval(() => {
				const now = dayjs()
				const timeLeft = totpExpiresAt.diff(now, 'seconds')

				if (timeLeft === 0) {
					clearInterval(interval)
					setTotpCountdown(0)
					return
				}

				setTotpCountdown(totpExpiresAt.diff(now, 'seconds'))
			}, 16.67)

			return () => clearInterval(interval)
		},
		[totpCountdown, totpExpiresAt],
	)

	return { totp, totpCountdown }
}
