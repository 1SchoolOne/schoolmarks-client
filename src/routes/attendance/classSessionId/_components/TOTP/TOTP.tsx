import { Col, Flex, Progress, Row, Typography } from 'antd'

import { useController } from './TOTP-controller'

interface TOTPCountdownProps {
	totp: string | undefined
	totpCountdown: number
}

function TOTP() {
	const { totp, totpCountdown } = useController()

	return <TOTPCountdown totp={totp} totpCountdown={totpCountdown} />
}

function TOTPCountdown(props: TOTPCountdownProps) {
	const { totp, totpCountdown } = props

	return (
		<Row className="totp-container" gutter={[8, 8]}>
			<Col span={24}>
				<Flex align="center" justify="center" gap={6}>
					<Typography.Title level={3}>Code : {totp ? totp : '-'}</Typography.Title>
				</Flex>
			</Col>
			{totp && (
				<Col span={24}>
					<Progress
						data-testid="totp-countdown-id"
						percent={(totpCountdown / 15) * 100 + 1}
						format={() => ''}
						strokeColor={totpCountdown > 5 ? 'var(--ant-color-info)' : 'var(--ant-color-error)'}
					/>
				</Col>
			)}
		</Row>
	)
}

TOTP.Countdown = TOTPCountdown

export { TOTP }
