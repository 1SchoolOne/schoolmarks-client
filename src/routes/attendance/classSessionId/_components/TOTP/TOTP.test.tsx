import { screen } from '@testing-library/react'

import { renderWithProviders } from '@helpers/test'

import { getMockServer } from '../../../../../../tests/mockServer'
import { TOTP } from './TOTP'

const mockServer = getMockServer('admin')

describe('<TOTP.Countdown />', () => {
	beforeAll(() => mockServer.listen({ onUnhandledRequest: 'warn' }))
	beforeEach(() => mockServer.resetHandlers())
	afterAll(() => mockServer.close())

	it('conforme au snapshot avec TOTP', async () => {
		const { asFragment } = renderWithProviders(<TOTP.Countdown totp="123456" totpCountdown={15} />)

		expect(asFragment()).toMatchSnapshot()
	})

	it('conforme au snapshot sans TOTP', async () => {
		const { asFragment } = renderWithProviders(
			<TOTP.Countdown totp={undefined} totpCountdown={15} />,
		)

		expect(asFragment()).toMatchSnapshot()
	})

	it('affiche le TOTP', async () => {
		renderWithProviders(<TOTP.Countdown totp="123456" totpCountdown={15} />)

		expect(screen.getByTestId('totp-countdown-id')).toBeDefined()
		expect(screen.getByText('Code : 123456')).toBeDefined()
	})

	it("affiche un '-' s'il n'y a pas de TOTP", async () => {
		renderWithProviders(<TOTP.Countdown totp={undefined} totpCountdown={15} />)

		expect(screen.queryByTestId('totp-countdown-id')).toBeNull()
		expect(screen.getByText('Code : -')).toBeDefined()
	})
})
