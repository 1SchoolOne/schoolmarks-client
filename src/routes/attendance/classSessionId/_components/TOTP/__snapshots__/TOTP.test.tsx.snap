// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TOTP.Countdown /> > conforme au snapshot avec TOTP 1`] = `
<DocumentFragment>
  <div
    class="ant-row totp-container css-dev-only-do-not-override-yta69t css-var-r0"
    style="margin-left: -4px; margin-right: -4px; row-gap: 8px;"
  >
    <div
      class="ant-col ant-col-24 css-dev-only-do-not-override-yta69t css-var-r0"
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        class="ant-flex css-dev-only-do-not-override-yta69t css-var-r0 ant-flex-align-center ant-flex-justify-center"
        style="gap: 6px;"
      >
        <h3
          class="ant-typography css-dev-only-do-not-override-yta69t css-var-r0"
        >
          Code : 123456
        </h3>
      </div>
    </div>
    <div
      class="ant-col ant-col-24 css-dev-only-do-not-override-yta69t css-var-r0"
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="101"
        class="ant-progress ant-progress-status-success ant-progress-line ant-progress-line-align-end ant-progress-line-position-outer ant-progress-show-info ant-progress-default css-dev-only-do-not-override-yta69t css-var-r0"
        data-testid="totp-countdown-id"
        role="progressbar"
      >
        <div
          class="ant-progress-outer"
          style="width: 100%;"
        >
          <div
            class="ant-progress-inner"
          >
            <div
              class="ant-progress-bg ant-progress-bg-outer"
              style="width: 100%; height: 8px; --progress-line-stroke-color: var(--ant-color-info); --progress-percent: 1;"
            />
          </div>
          <span
            class="ant-progress-text ant-progress-text-end ant-progress-text-outer"
            title=""
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TOTP.Countdown /> > conforme au snapshot sans TOTP 1`] = `
<DocumentFragment>
  <div
    class="ant-row totp-container css-dev-only-do-not-override-yta69t css-var-r1"
    style="margin-left: -4px; margin-right: -4px; row-gap: 8px;"
  >
    <div
      class="ant-col ant-col-24 css-dev-only-do-not-override-yta69t css-var-r1"
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        class="ant-flex css-dev-only-do-not-override-yta69t css-var-r1 ant-flex-align-center ant-flex-justify-center"
        style="gap: 6px;"
      >
        <h3
          class="ant-typography css-dev-only-do-not-override-yta69t css-var-r1"
        >
          Code : -
        </h3>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
